design-proposals/klip-24-key-column-semantics-in-queries.md

```diff
@@ -0,0 +1,574 @@
+to ensure the match the new syntax.
+
+## Compatibility Implications
+
+All preexisting queries, i.e. those with plans serialized to the command topic, will continue to
+work and we have extensive tests covering this.
+
+Some existing SQL, if reissued, will fail if a persistent query's projection does not include the
+key column. However, a helpful error message will inform the user of the changes they need to make
+to resolve this.  Resolution is simple: just add the key column to the projection!
+
+## Security Implications
+
+None.
+
+## Rejected alternatives
+
+### Magic key column copying if no key column in persistent query's projection.
+
+Design as above, but not _requiring_ the key column in the projection. Instead, allow the key
+column in the projection and implicitly copy it across if its not there.  This was rejected for two
+key reasons:
+  a. Potentially confusing 'magic implicits' - the output contains columns the projection doesn't
+     specify.
+  b. It over complicates the implementation.
+
+### Full support for no key column in persistent query's projection.
+
+Design as above, but allowing persistent query projections to not include the key column in the
+projection. If not present, the created data source would have no _exposed_ key column.
+
+Given that the created source is actually a materialized view, it seems completely reasonable to
+allow users to control the set of columns the view exposes. Any requirement internally for the key,
+e.g. to allow updates to be processed correctly, would remain. The key would just not be available
+in downstream queries.
+
+This was rejected for this KLIP as it would involve considerably more work. This may be picked up in
+a future KLIP.
+
+### Storing all join columns in the key of the result
+
+A more correct solution for handling columns within a join may look to store all join columns in the
+Kafka record's key, for example:
+
+```sql
+SELECT * FROM INPUT I1 JOIN INPUT_2 I2 ON I1.ID = I2.ID;
+-- resulting columns: I1_ID INT, I1_V0 INT, I1_V1 INT, I2_ID INT, I2_V0 INT, I2_V1 INT
+
+CREATE TABLE OUTPUT AS
+   SELECT * FROM INPUT I1 JOIN INPUT_2 I2 ON I1.ID = I2.ID;
+-- resulting schema: I1_ID INT KEY, I2_ID INT KEY, I1_V0 INT, I1_V1 INT, I2_V0 INT, I2_V1 INT
+```
+
+Note that both `I1_ID` and `I2_ID` are marked as key columns. Such an approach may be required if
+ksqlDB is to support join criterion other than the current equality.
+
+However, this is rejected as a solution for now for the following reasons:
+  a. Such a solution requires ksqlDB to support multiple key columns. It currently does not, and this
+     KLIP is part of the work moving towards such support. Hence its a chicken and egg problem.
+  b. Such a solution requires Streams to be able to correctly handle the multiple key columns
+     correctly, which it currently does not.  This is particularly challenging for outer joins,
```

Comment:
typo: `to correctly handle [...] correctly,`

> which it currently does not

And maybe never will. However, as discussed in person, there seems to be a solution by compiling down some additional logic into the Kafka Streams program.

Reply:
Updated text.  Rest is ... tbd.